tests/metadata/test_metadata_api.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
+
+# (C) Copyright 2020 ECMWF.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+#
+
+import numpy as np
+import pytest
+
+from earthkit.data import from_object, from_source
+from earthkit.data.testing import earthkit_examples_file
+
+try:
+    from earthkit.data.metadata import Metadata
+except ImportError:
+    Metadata = None
+
+try:
+    from earthkit.data.metadata import RawMetadata
+except ImportError:
+    RawMetadata = None
+
+try:
+    from earthkit.data.readers.grib.metadata import GribMetadata
+except ImportError:
+    GribMetadata = None
+
+
+def test_abstract_metadata():
+    # Metadata properties:
+    # * immutable
+    # * behaves like a dict
+    # * can be updated (creating a new object, either a copy or a view)
+    # Updates can be made with compatible subclasses, e.g. update a GribMetadata with a RawMetadata
+    assert Metadata is not None
+
+
+def test_raw_metadata():
+    assert RawMetadata is not None
+    assert issubclass(RawMetadata, Metadata)
+
+    md = RawMetadata({"shortName": "2t", "perturbationNumber": 5})
+    assert md["shortName"] == "2t"
+    assert md["perturbationNumber"] == 5
+    assert md.get("shortName") == "2t"
+    with pytest.raises(KeyError):
+        md["nonExistentKey"]
+    assert md.get("nonExistentKey", 12) == 12
```

Comment:
What should happen for:

```python
md.get("nonExistentKey")
```
Should we raise `KeyError` or return None as a `dict` would do?

Reply:
I would return None, to match `dict` and the behaviour of the default value when given

Reply:
Thank you @oiffrig! There is another question related to the type cast. On a `Field` we can call `metadata()` like this to specify the value type we want:

```python
f.metadata("centre", astype=str)
f.metadata("centre", astype=int)
f.metadata("centre:s")
f.metadata("centre:str")
f.metadata("centre:l")
f.metadata("centre:int")
```
How would the Metadata object support these? The last 4 calls are of course supported for GRIB by `[]` and `get()`.

Reply:
Not totally sure. What I would suggest is to have that as an overridable behaviour: the metadata object knows a few types and can convert from them (raising an error if the type is not supported), and specific implementations (e.g. GRIB) can extend the default behaviour by using functionalities of the underlying backend

Reply:
Thanks. Would that mean that `get` should work like that:

```python
md.get("centre", astype=int)
```